File: Semester-2.1/Notes/CSE 206/Stack.py
class Stack:
    __maxSize = 0
    __a = []
    __top = -1

    def __init__(self, size: int):
        self.__maxSize = size

    def size(self) -> int:
        return self.__top + 1

    def isFull(self) -> bool:
        return self.__top == self.__maxSize - 1

    def isEmpty(self) -> bool:
        return self.__top == - 1

    def top(self) -> int:
        if self.__top != -1:
            return self.__a[self.__top]
        else:
            return -1

    def isPeek(self, n: int) -> int:
        return n == self.top()

    def push(self, n: int) -> int:
        if not self.isFull():
            self.__a.append(n)
            self.__top += 1
        else:
            print("Stack overflow!\n")

    def pop(self) -> int:
        if not self.isEmpty():
            item = self.__a[self.__top]
            self.__top -= 1
            self.__a.pop()
            return item
        else:
            print("Stack underflow!\n")
            return -1

    def printAll(self):
        for n in self.__a:
            print(n, end=" ")
        print()


size = int(input("Enter Stack capcity: "))
s = Stack(size)
print()

while (True):
    print("Stack Operation:\n1. Push item\n2. Pop item\n3. View Top\n4. Check if the stack is full?\n5. Check if the stack is empty?\n6. Print all elements.\n0. Exit")
    option = int(input("Select: "))
    print()

    match option:
        case 1:
            n = int(input("Enter a number: "))
            s.push(n)

        case 2:
            item = s.pop()
            if item != -1:
                print(f"{item} has been poped.\n")

        case 3:
            if s.top() != -1:
                print(f"Top item: {s.top()}.\n")
            else:
                print("No element in Stack.\n")

        case 4:
            if s.isFull():
                print("Stack is full.\n")
            else:
                print("Stack is not full.\n")

        case 5:
            if s.isEmpty():
                print("Stack is empty.\n")
            else:
                print("Stack is not empty.\n")

        case 6:
            print("Stack items: ")
            s.printAll()

        case 0:
            break

File: Semester-2.1/Notes/CSE 206/Stack.cpp
#include <bits/stdc++.h>
using namespace std;

class Stack
{
private:
   int *a;
   int capacity = 10;
   int top = -1;

public:
   Stack(int _size_)
   {
      capacity = _size_;
      a = new int[_size_];
   }

   int size()
   {
      return top + 1;
   }

   bool isFull()
   {
      return top == capacity - 1;
   }

   bool isEmpty()
   {
      return top == -1;
   }

   int peek()
   {
      return a[top];
   }

   bool isPeek(int n)
   {
      return a[top] == n;
   }

   int push(int n)
   {
      if (!isFull())
      {
         top += 1;
         a[top] = n;
         return a[top];
      }
      else
      {
         cout << "Stack overflow!" << endl;
         return -1;
      }
   }

   int pop()
   {
      if (!isEmpty())
      {
         int item = a[top];
         top -= 1;
         return item;
      }
      else
      {
         cout << "Stack underflow!" << endl;
         return -1;
      }
   }

   void printAll()
   {
      if (!isEmpty())
      {
         cout << "\nStack items:" << endl;
         for (int i = 0; i <= top; i++)
            cout << *(a + i) << " ";
         cout << endl;
      }
      else
         cout << "No items left in the stack!" << endl;
   }
};

int main()
{
   int sz, n, item, option;

   cout << "Enter stack size: ";
   cin >> sz;
   Stack st(sz);

   while (true)
   {
      cout << "\nStack operation:\n1. Push item\n2. Pop item\n3. View Top\n4. Check if the stack is full?\n5. Check if the stack is empty?\n0. Exit" << endl;
      cout << "select: ";
      cin >> option;

      cout << endl;

      switch (option)
      {
      case 1:
         cout << "Enter a number: ";
         cin >> n;
         st.push(n);
         st.printAll();
         break;

      case 2:
         st.pop();
         st.printAll();
         break;

      case 3:
         if (st.peek() != -1)
            cout << "Top item: " << st.peek() << endl;
         else
            cout << "No element in Stack." << endl;
         break;

      case 4:
         if (st.isFull())
            cout << "Stack is full." << endl;
         else
            cout << "Stack is not full." << endl;
         break;

      case 5:
         if (st.isEmpty())
            cout << "Stack is empty." << endl;
         else
            cout << "Stack is not empty." << endl;
         break;

      case 0:
         cout << "Program terminated!" << endl;
         return 0;
      }
   }

   return 0;
}